src/crystal/system/unix/env.cr

```diff
@@ -74,29 +93,29 @@ module Crystal::System::Env
   # buffer and `envp` upfront to reduce overall allocations.
   def self.make_envp(env, clear_env) : LibC::Char**
     # When there are no adjustments in `env`, we can take a short cut and return
-    # an empty pointer or the current environment.
-    if env.nil? || env.empty?
-      return clear_env ? Pointer(LibC::Char*).malloc(1) : LibC.environ
+    # an empty pointer.
+    if clear_env && (env.nil? || env.empty?)
+      return Pointer(LibC::Char*).malloc(1)
     end
 
     envp = Array(LibC::Char*).new
 
     unless clear_env
-      Env.each do |key, value|
-        # Skip overrides in `env`
-        next if env.has_key?(key)
+      # Dup LibC.environ, skipping overrides in env.
+      each do |key, value|
```

Comment:
This is expensive and since we can't return `environ` anymore, this happens all the time.

I have a simple patch that yields slices to the kv pointer instead of strings, which avoids the intermediary string allocations and directly builds the "key=value" string. I'll push it in a follow up.